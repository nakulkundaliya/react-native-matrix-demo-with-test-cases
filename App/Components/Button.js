import React, { Component } from 'react';
// import PropTypes from 'prop-types';
import { TouchableOpacity, Text } from 'react-native';
import styles from './Styles/ButtonStyle';

const Button = props => {
  const { name, onPress, style, disabled } = props;
  return (
    <TouchableOpacity
      style={[styles.buttonStyle, style]}
      onPress={() => onPress()}
      disabled={disabled}
    >
      <Text style={styles.buttonTextStyle}>{name}</Text>
    </TouchableOpacity>
  );
};
export default Button;
