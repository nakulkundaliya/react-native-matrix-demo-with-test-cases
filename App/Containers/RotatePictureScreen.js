import React, { Component } from 'react';
import { ScrollView, TextInput, View, Text } from 'react-native';
import { connect } from 'react-redux';
import Button from '../Components/Button';
import Box from '../Components/Box';
import styles from './Styles/RotatePictureScreenStyle';
import { TouchableOpacity } from 'react-native-gesture-handler';
class RotatePictureScreen extends Component {
  static navigationOptions = {
    title: 'Rotate Picture'
  };

  constructor(props) {
    super(props);
    this.state = {
      n: null,
      nMatrix: null,
      // matrix: [[1, 2, 3], [1, 2, 3], [1, 2, 3]]
      matrix: new Array(3)
    };
  }

  addMatrix = () => {
    const { n } = this.state;
    if (isNaN(n) || n < 2) {
      this.setState({ nMarix: null });
      alert('Please insert proper value');
    } else {
      let generatArray = JSON.parse(
        JSON.stringify(Array(parseInt(n)).fill(Array(parseInt(n)).fill('')))
      );

      for (var i = 0; i < parseInt(n); i++) {
        for (var j = 0; j < parseInt(n); j++) {
          generatArray[i][j] = Math.floor(Math.random() * 255) + 0;
        }
      }
      this.setState({ nMatrix: n, matrix: generatArray });
    }
  };

  rotateImage = () => {
    const { matrix } = this.state;
    const n = matrix.length;
    const x = Math.floor(n / 2);
    const y = n - 1;
    for (let i = 0; i < x; i++) {
      for (let j = i; j < y - i; j++) {
        k = matrix[i][j];
        matrix[i][j] = matrix[y - j][i];
        matrix[y - j][i] = matrix[y - i][y - j];
        matrix[y - i][y - j] = matrix[j][y - i];
        matrix[j][y - i] = k;
      }
    }
    this.setState({ matrix });
  };

  renderAddMatrixView() {
    const { n } = this.state;
    return (
      <View style={styles.addMatrixViewStyle}>
        <TextInput
          placeholder="n value"
          keyboardType="number-pad"
          style={styles.textInputStyle}
          onChangeText={text => this.setState({ n: text })}
          value={n}
          maxLength={1}
        />
        <Button
          name="Add"
          onPress={this.addMatrix}
          style={styles.addButtonStyle}
          disabled={!n}
        />
      </View>
    );
  }
  renderNote() {
    return <Text style={styles.noteTextStyle}>Note : Please add n value </Text>;
  }

  renderMatrix() {
    const { nMatrix, matrix } = this.state;
    return (
      <View style={styles.matrixViewStyle}>
        {matrix.map(row => {
          return (
            <View style={styles.matrixRowStyle}>
              {row.map(column => {
                return <Box value={column} nMatrix={nMatrix} />;
              })}
            </View>
          );
        })}
      </View>
    );
  }

  renderRotateImage() {
    return (
      <Button
        name="Rotate"
        onPress={this.rotateImage}
        style={styles.rotateImageButtonStyle}
      />
    );
  }

  render() {
    const { nMatrix } = this.state;
    return (
      <ScrollView style={styles.container}>
        {this.renderAddMatrixView()}
        {this.renderNote()}
        {nMatrix && this.renderMatrix()}
        {nMatrix && this.renderRotateImage()}
      </ScrollView>
    );
  }
}

const mapStateToProps = state => {
  return {};
};

const mapDispatchToProps = dispatch => {
  return {};
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(RotatePictureScreen);
