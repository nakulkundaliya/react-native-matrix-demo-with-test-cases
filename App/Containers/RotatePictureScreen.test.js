import React from 'react';
import { shallow } from 'enzyme';
import RotatePictureScreen from './RotatePictureScreen';
import Button from '../Components/Button';

describe('RotatePictureScreen', () => {
  describe('Rendering', () => {
    let component;
    beforeEach(() => {
      component = shallow(<RotatePictureScreen />);
    });

    it('should be defined', () => {
      expect(RotatePictureScreen).toBeDefined();
    });

    it('should match to snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    it('should check form validation', () => {
      component.setState({ n: '' });
      alert = jest.fn();
      const instance = component.instance();
      jest.spyOn(instance, 'addMatrix');
      component
        .find(Button)
        .first()
        .props()
        .onPress();
      expect(alert).toBeCalled();
    });

    it('should generate n metrix', () => {
      component.setState({ n: '3' });
      const instance = component.instance();
      jest.spyOn(instance, 'addMatrix');
      component
        .find(Button)
        .first()
        .props()
        .onPress();
      expect(instance.state.nMatrix).toBe('3');
      expect(instance.state.matrix.length).toBe(3);
    });

    it('should rotate metrix', () => {
      component.setState({ n: '3' });
      const instance = component.instance();
      jest.spyOn(instance, 'addMatrix');
      component
        .find(Button)
        .first()
        .props()
        .onPress();
      jest.spyOn(instance, 'rotateImage');

      let generatedArray = instance.state.matrix;
      let firstIndexValue = generatedArray[0][0];
      component
        .find(Button)
        .last()
        .props()
        .onPress();
      let rotateArray = instance.state.matrix;
      let newIndexValue = rotateArray[0][2];

      expect(firstIndexValue).toBe(newIndexValue);
    });
  });
});
