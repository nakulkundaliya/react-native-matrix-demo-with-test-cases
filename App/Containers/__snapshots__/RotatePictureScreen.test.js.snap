// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RotatePictureScreen Rendering should match to snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RotatePictureScreen />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <View
          style={
            Object {
              "alignSelf": "center",
              "flex": 1,
              "flexDirection": "row",
              "marginVertical": 20,
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            keyboardType="number-pad"
            maxLength={1}
            onChangeText={[Function]}
            placeholder="n value"
            rejectResponderTermination={true}
            style={
              Object {
                "borderColor": "#595959",
                "borderRadius": 10,
                "borderWidth": 1,
                "height": 50,
                "justifyContent": "center",
                "paddingHorizontal": 10,
                "width": 150,
              }
            }
            underlineColorAndroid="transparent"
            value={null}
          />
          <Button
            disabled={true}
            name="Add"
            onPress={[Function]}
            style={
              Object {
                "marginLeft": 10,
                "marginVertical": 0,
                "width": 150,
              }
            }
          />
        </View>,
        <Text
          style={
            Object {
              "fontStyle": "italic",
              "marginTop": 10,
              "textAlign": "center",
            }
          }
        >
          Note : Please add n value 
        </Text>,
        null,
        null,
      ],
      "style": Object {
        "backgroundColor": "rgba(0,0,0,0)",
        "flex": 1,
        "marginHorizontal": 15,
        "paddingTop": 10,
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <TextInput
              allowFontScaling={true}
              keyboardType="number-pad"
              maxLength={1}
              onChangeText={[Function]}
              placeholder="n value"
              rejectResponderTermination={true}
              style={
                Object {
                  "borderColor": "#595959",
                  "borderRadius": 10,
                  "borderWidth": 1,
                  "height": 50,
                  "justifyContent": "center",
                  "paddingHorizontal": 10,
                  "width": 150,
                }
              }
              underlineColorAndroid="transparent"
              value={null}
            />,
            <Button
              disabled={true}
              name="Add"
              onPress={[Function]}
              style={
                Object {
                  "marginLeft": 10,
                  "marginVertical": 0,
                  "width": 150,
                }
              }
            />,
          ],
          "style": Object {
            "alignSelf": "center",
            "flex": 1,
            "flexDirection": "row",
            "marginVertical": 20,
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "allowFontScaling": true,
              "keyboardType": "number-pad",
              "maxLength": 1,
              "onChangeText": [Function],
              "placeholder": "n value",
              "rejectResponderTermination": true,
              "style": Object {
                "borderColor": "#595959",
                "borderRadius": 10,
                "borderWidth": 1,
                "height": 50,
                "justifyContent": "center",
                "paddingHorizontal": 10,
                "width": 150,
              },
              "underlineColorAndroid": "transparent",
              "value": null,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "disabled": true,
              "name": "Add",
              "onPress": [Function],
              "style": Object {
                "marginLeft": 10,
                "marginVertical": 0,
                "width": 150,
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Note : Please add n value ",
          "style": Object {
            "fontStyle": "italic",
            "marginTop": 10,
            "textAlign": "center",
          },
        },
        "ref": null,
        "rendered": "Note : Please add n value ",
        "type": [Function],
      },
      null,
      null,
    ],
    "type": [MockFunction],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <View
            style={
              Object {
                "alignSelf": "center",
                "flex": 1,
                "flexDirection": "row",
                "marginVertical": 20,
              }
            }
          >
            <TextInput
              allowFontScaling={true}
              keyboardType="number-pad"
              maxLength={1}
              onChangeText={[Function]}
              placeholder="n value"
              rejectResponderTermination={true}
              style={
                Object {
                  "borderColor": "#595959",
                  "borderRadius": 10,
                  "borderWidth": 1,
                  "height": 50,
                  "justifyContent": "center",
                  "paddingHorizontal": 10,
                  "width": 150,
                }
              }
              underlineColorAndroid="transparent"
              value={null}
            />
            <Button
              disabled={true}
              name="Add"
              onPress={[Function]}
              style={
                Object {
                  "marginLeft": 10,
                  "marginVertical": 0,
                  "width": 150,
                }
              }
            />
          </View>,
          <Text
            style={
              Object {
                "fontStyle": "italic",
                "marginTop": 10,
                "textAlign": "center",
              }
            }
          >
            Note : Please add n value 
          </Text>,
          null,
          null,
        ],
        "style": Object {
          "backgroundColor": "rgba(0,0,0,0)",
          "flex": 1,
          "marginHorizontal": 15,
          "paddingTop": 10,
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <TextInput
                allowFontScaling={true}
                keyboardType="number-pad"
                maxLength={1}
                onChangeText={[Function]}
                placeholder="n value"
                rejectResponderTermination={true}
                style={
                  Object {
                    "borderColor": "#595959",
                    "borderRadius": 10,
                    "borderWidth": 1,
                    "height": 50,
                    "justifyContent": "center",
                    "paddingHorizontal": 10,
                    "width": 150,
                  }
                }
                underlineColorAndroid="transparent"
                value={null}
              />,
              <Button
                disabled={true}
                name="Add"
                onPress={[Function]}
                style={
                  Object {
                    "marginLeft": 10,
                    "marginVertical": 0,
                    "width": 150,
                  }
                }
              />,
            ],
            "style": Object {
              "alignSelf": "center",
              "flex": 1,
              "flexDirection": "row",
              "marginVertical": 20,
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "allowFontScaling": true,
                "keyboardType": "number-pad",
                "maxLength": 1,
                "onChangeText": [Function],
                "placeholder": "n value",
                "rejectResponderTermination": true,
                "style": Object {
                  "borderColor": "#595959",
                  "borderRadius": 10,
                  "borderWidth": 1,
                  "height": 50,
                  "justifyContent": "center",
                  "paddingHorizontal": 10,
                  "width": 150,
                },
                "underlineColorAndroid": "transparent",
                "value": null,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "disabled": true,
                "name": "Add",
                "onPress": [Function],
                "style": Object {
                  "marginLeft": 10,
                  "marginVertical": 0,
                  "width": 150,
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Note : Please add n value ",
            "style": Object {
              "fontStyle": "italic",
              "marginTop": 10,
              "textAlign": "center",
            },
          },
          "ref": null,
          "rendered": "Note : Please add n value ",
          "type": [Function],
        },
        null,
        null,
      ],
      "type": [MockFunction],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
